Strsplittrail

brodieG opened the pull request:
Proposed "fix" for #26 (Fixes #26).  A few noteworthy comments:

* I "fixed" this issue only in `col_strsplit`.  It could potentially be "fixed" in `crayon:::re_table` which would be cleaner, but would affect multiple other functions
* I also enabled the `empty` parameter for `non_matching`; since it wasn't being used (I think), I updated all calls to that function to specify `empty=TRUE` to avoid changing behavior

Let me know if you want me to change / revert any of this.

brodieG commented:
Fyi, I'll submit the resolved conflicts later today.

brodieG commented:
@gaborcsardi, fyi, the branch is now up to date.

brodieG commented:
@gaborcsardi, let me know if you intend to eventually merge this one or not.  Fine if you don't, but it would be helpful for me to know one way or other since it affects how I write my package that imports crayon.

gaborcsardi commented on R/string_operations.r:
```diff
@@ -212,6 +214,14 @@ col_strsplit <- function(x, split, ...) {
+  chunks <- lapply(
+    chunks,
+    function(y)
+      if(nrow(y) > 1L && !tail(y, 1L)[, "length"]) head(y, -1L) else y
```
Why the `nrow(y) > 1L` condition? I have 
```r
❯ strsplit("", "a")[[1]]
character(0)
```
so we want to drop a single `""` as well, I think.

brodieG replied:
I'll have to look at it again later.  Clearly this is wrong, but I think I was trying to ensure that we don't drop this:

```
> strsplit("a", "a")[[1]]
[1] ""
```
Hopefully I can look at it this evening.

brodieG commented:
Okay, so unfortunately addressing the corner case you highlight is not completely straightforward.  This latest push is *NOT* ready for merging as there remains some corner cases and I want to test this more thoroughly.

I put it here for your inspection in because the (currently partial) solution I put in is not particularly elegant.  As far as I can tell, I can't easily rely on the existing infrastructure to handle the corner cases, as passing 0-row chunks causes `col_substr` and co. to break.  Instead, I have to handle the corner cases in `col_strsplit` directly.  I could probably fix the internal functions to be more elegant, but that will likely mean modifying a fair bit more code.

Let me know if have any concerns with me just handling the corner cases as the `col_strsplit`.

gaborcsardi commented:
Thanks! I'll try to look at it this evening.

brodieG commented:
Okay, this should be good to go now, assuming you're okay with how I've addressed the issues.  I wish `strsplit` was a bit more consistent in how it handles corner cases.

brodieG commented on tests/testthat/test-operations.R:
```diff
@@ -119,8 +128,30 @@ test_that("col_strsplit multiple strings", {
+  )
+})
 
-  expect_equal(col_strsplit("", "-"), list(""))
```
Note that these tests no longer behave like they used to so I changed them.

brodieG commented:
I addressed the last remaining issue I opened.  Note that one issue is that `substr("abc", "hello", "hello")` works (returns NA), whereas I just sidestepped that by disallowing string position arguments in `col_substr`.

Also, appveyor is failing, but I think the issue is not related to the code changes I made (for some reason `devtools` didn't load properly).

Finally, haven't heard much one way or another on this pull request.  Feedback would be appreciated.  Thanks.

gaborcsardi commented:
I am sorry this is taking long. I'll look at it today.

brodieG commented:
Let me know if you want me to simplify this pull request by breaking it up into smaller ones.

gaborcsardi commented:
Thanks much! I am sorry that it took so long!

brodieG commented:
Great, thanks for accepting the pull.
